zetaclient/chains/bitcoin/observer/event.go

```diff
@@ -0,0 +1,253 @@
+
+	// InboundProcessabilityComplianceViolation represents a compliance violation
+	InboundProcessabilityComplianceViolation
+)
+
+// BTCInboundEvent represents an incoming transaction event
+type BTCInboundEvent struct {
+	// FromAddress is the first input address
+	FromAddress string
+
+	// ToAddress is the ZEVM receiver address
+	ToAddress string
+
+	// Value is the amount of BTC
+	Value float64
+
+	// DepositorFee is the deposit fee
+	DepositorFee float64
+
+	// MemoBytes is the memo of inbound
+	MemoBytes []byte
+
+	// MemoStd is the standard inbound memo if it can be decoded
+	MemoStd *memo.InboundMemo
+
+	// BlockNumber is the block number of the inbound
+	BlockNumber uint64
+
+	// TxHash is the hash of the inbound
+	TxHash string
+}
+
+// Processability returns the processability of the inbound event
+func (event BTCInboundEvent) Processability() InboundProcessability {
+	// compliance check on sender and receiver addresses
+	if config.ContainRestrictedAddress(event.FromAddress, event.ToAddress) {
+		return InboundProcessabilityComplianceViolation
+	}
+
+	// compliance check on receiver, revert/abort addresses in standard memo
+	if event.MemoStd != nil {
+		if config.ContainRestrictedAddress(
+			event.MemoStd.Receiver.Hex(),
+			event.MemoStd.RevertOptions.RevertAddress,
+			event.MemoStd.RevertOptions.AbortAddress,
+		) {
+			return InboundProcessabilityComplianceViolation
+		}
+	}
+
+	// donation check
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
+		return InboundProcessabilityDonation
+	}
+
+	return InboundProcessabilityGood
+}
+
+// DecodeMemoBytes decodes the contained memo bytes as either standard or legacy memo
+func (event *BTCInboundEvent) DecodeMemoBytes(chainID int64) error {
+	var (
+		err            error
+		isStandardMemo bool
+		memoStd        *memo.InboundMemo
+		receiver       ethcommon.Address
+	)
+
+	// skip decoding donation tx as it won't go through zetacore
+	if bytes.Equal(event.MemoBytes, []byte(constant.DonationMessage)) {
+		return nil
+	}
+
+	// try to decode the standard memo as the preferred format
+	// the standard memo is NOT enabled for Bitcoin mainnet
+
+	if chainID != chains.BitcoinMainnet.ChainId {
+		memoStd, isStandardMemo, err = memo.DecodeFromBytes(event.MemoBytes)
+	}
+
+	// process standard memo or fallback to legacy memo
+	if isStandardMemo {
+		// skip standard memo that carries improper data
+		if err != nil {
+			return errors.Wrap(err, "standard memo contains improper data")
+		}
+
+		// validate the content of the standard memo
+		err = ValidateStandardMemo(*memoStd, chainID)
+		if err != nil {
+			return errors.Wrap(err, "invalid standard memo for bitcoin")
+		}
+
+		event.MemoStd = memoStd
+		receiver = memoStd.Receiver
+	} else {
+		parsedAddress, _, err := memo.DecodeLegacyMemoHex(hex.EncodeToString(event.MemoBytes))
+		if err != nil { // unreachable code
+			return errors.Wrap(err, "invalid legacy memo")
+		}
+		receiver = parsedAddress
```

Comment:
If it's unreachable couldn't we simply do:
```suggestion
		receiver, _, _ = memo.DecodeLegacyMemoHex(hex.EncodeToString(event.MemoBytes))
```

Reply:
I see what you mean. I won't assume that the implementation of `DecodeLegacyMemoHex` will never change in future.